pkg/services/pluginsintegration/plugincontext/key.go

```diff
@@ -0,0 +1,27 @@
+package plugincontext
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+)
+
+type KeyProvider interface {
+	DataSourceKey(ctx context.Context, dsSettings *backend.DataSourceInstanceSettings) string
+	AppKey(ctx context.Context, pluginID string, orgID int64) string
+}
+
+func ProvideKeyService() *KeyService {
+	return &KeyService{}
+}
+
+type KeyService struct{}
+
+func (s *KeyService) DataSourceKey(_ context.Context, dsSettings *backend.DataSourceInstanceSettings) string {
```

Comment:
`DataSourceID`?  or is this meant to return something different? Some documentation would help

Reply:
Yeah sorry without docs or the right context this can be confusing. 

So right now in the SDK, when we cache either datasource or app "instances", the cache key is defined in the SDK:
- For [apps](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/app/instance_provider.go#L51)
- For [data sources](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/datasource/instance_provider.go#L48)

A new `pluginContext.Key` field can be used to replace this in the SDK so that Grafana can decide how the instances are cached. Since there is a separate key for Apps and DataSources, I'm keeping the same behaviour here by using this `KeyProvider` to be used as part of `plugincontext.Provider`. Does that make sense? 🤔

Reply:
yes :+1: okay that makes sense. The only thing to have into account is plugins using an old Grafana version with a new SDK version (but we can add a fallback path if `pluginContext.Key` is empty).